types/request.d.ts

```diff
@@ -38,7 +38,35 @@ export interface RequestRouteOptions<ContextConfig = ContextConfigDefault, Schem
  * FastifyRequest is an instance of the standard http or http2 request objects.
  * It defaults to http.IncomingMessage, and it also extends the relative request object.
  */
-export interface FastifyRequest<RouteGeneric extends RouteGenericInterface = RouteGenericInterface,
+export type FastifyRequest<
+  RouteGeneric extends RouteGenericInterface = RouteGenericInterface,
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestDefaultExpression<RawServer> = RawRequestDefaultExpression<RawServer>,
+  SchemaCompiler extends FastifySchema = FastifySchema,
+  TypeProvider extends FastifyTypeProvider = FastifyTypeProviderDefault,
+  ContextConfig = ContextConfigDefault,
+  Logger extends FastifyBaseLogger = FastifyBaseLogger,
+  Decorators extends FastifyDecorators['request'] = FastifyDecorators['request'],
+  RequestType extends FastifyRequestType = ResolveFastifyRequestType<TypeProvider, SchemaCompiler, RouteGeneric>
+// ^ Temporary Note: RequestType has been re-ordered to be the last argument in
+//   generic list. This generic argument is now considered optional as it can be
+//   automatically inferred from the SchemaCompiler, RouteGeneric and TypeProvider
+//   arguments. Implementations that already pass this argument can either omit
+//   the RequestType (preferred) or swap Logger and RequestType arguments when
+//   creating custom types of FastifyRequest. Related issue #4123
+> = Decorators & BaseFastifyRequest<
+  RouteGeneric,
+  RawServer,
+  RawRequest,
+  SchemaCompiler,
+  TypeProvider,
+  ContextConfig,
+  Logger,
+  RequestType
+>
+
+export interface BaseFastifyRequest<
```

Comment:
`FastifyRequest` might be imported and customized in many code bases... Are we sure this is a good idea to modify it? Shouldn't we create a new Type for the problem we are solving?

Reply:
FastifyRequest has the same interface, when looked from the outside. It was converted the way it was to incorporate Decorators. Previously, FastifyRequest was an interface, and i couldn't add a type from the generics to the interface. So a straightforward refactor was to introduce an intersection type.

From the outside, FastifyRequest shouldn't have changed when it came to properties it had, it should have only gotten a new prop. The change was meant to be backwards compatible. Do you have an example of breakage?

Reply:
Yep, maybe one: https://github.com/fastify/fastify/pull/5672/files/b2c0a9c2d49a288fce99c3f4726329f188ba6e2c#r1818060338